src/descent/halley.jl

```diff
@@ -0,0 +1,87 @@
+"""
+    HalleyDescent(; linsolve = nothing, precs = DEFAULT_PRECS)
+
+Compute the descent direction as ``J δu = -fu``. For non-square Jacobian problems, this is
+commonly referred to as the Gauss-Newton Descent.
+
+See also [`Dogleg`](@ref), [`SteepestDescent`](@ref), [`DampedNewtonDescent`](@ref).
+"""
+@kwdef @concrete struct HalleyDescent <: AbstractDescentAlgorithm
+    linsolve = nothing
+    precs = DEFAULT_PRECS
+end
+
+using TaylorDiff: derivative
+
+function Base.show(io::IO, d::HalleyDescent)
+    modifiers = String[]
+    d.linsolve !== nothing && push!(modifiers, "linsolve = $(d.linsolve)")
+    d.precs !== DEFAULT_PRECS && push!(modifiers, "precs = $(d.precs)")
+    print(io, "HalleyDescent($(join(modifiers, ", ")))")
+end
+
+supports_line_search(::HalleyDescent) = true
+
+@concrete mutable struct HalleyDescentCache{pre_inverted} <:
+                         AbstractDescentCache
+    f
+    p
+    δu
+    δus
+    b
+    lincache
+    timer
+end
+
+@internal_caches HalleyDescentCache :lincache
+
+function __internal_init(
+        prob::NonlinearProblem, alg::HalleyDescent, J, fu, u; shared::Val{N} = Val(1),
+        pre_inverted::Val{INV} = False, linsolve_kwargs = (;), abstol = nothing,
+        reltol = nothing, timer = get_timer_output(), kwargs...) where {INV, N}
+    @bb δu = similar(u)
+    @bb b = similar(u)
+    δus = N ≤ 1 ? nothing : map(2:N) do i
+        @bb δu_ = similar(u)
+    end
+    INV && return HalleyDescentCache{true}(prob.f, prob.p, δu, δus, b, nothing, timer)
+    lincache = LinearSolverCache(
+        alg, alg.linsolve, J, _vec(fu), _vec(u); abstol, reltol, linsolve_kwargs...)
+    return HalleyDescentCache{false}(prob.f, prob.p, δu, δus, b, lincache, timer)
+end
+
+function __internal_solve!(
+        cache::HalleyDescentCache{INV}, J, fu, u, idx::Val = Val(1);
+        skip_solve::Bool = false, new_jacobian::Bool = true, kwargs...) where {INV}
+    δu = get_du(cache, idx)
+    skip_solve && return δu, true, (;)
+    if INV
+        @assert J!==nothing "`J` must be provided when `pre_inverted = Val(true)`."
+        @bb δu = J × vec(fu)
+    else
+        @static_timeit cache.timer "linear solve 1" begin
+            δu = cache.lincache(;
+                A = J, b = _vec(fu), kwargs..., linu = _vec(δu), du = _vec(δu),
+                reuse_A_if_factorization = !new_jacobian || (idx !== Val(1)))
+            δu = _restructure(get_du(cache, idx), δu)
+        end
+    end
+    b = cache.b
+    # compute the hessian-vector-vector product
+    hvvp = derivative(x -> cache.f(x, cache.p), u, δu, 2)
```

Comment:
```suggestion
    hvvp = derivative(Base.Fix2(cache.f, cache.p), u, δu, 2)
```

Reply:
This will also need a case for inplace `cache.f`

Reply:
Got it. TaylorDiff.jl haven't got a in-place mode yet, but will do this week